### PR TITLE
Listen issuance events

### PR DESCRIPTION
**Pull request by leonprou**
## Proposed changes

- Adding event listeners to db.
- Saving all communities on server side

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: server/services/events.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  events.forEach((event) => eventCallback(error, event))
+}
+
+CurrencyFactoryContract.events.TokenCreated(eventCallback)
+
+const getPastEvents = async () => {
+  const lastBlockNumber = await communityUtils.getLastBlockNumber()
+  console.log(lastBlockNumber)
+  CurrencyFactoryContract.getPastEvents('TokenCreated', {fromBlock: lastBlockNumber, toBlock: 'latest'}, eventsCallback)
```
Comment (bejavu) [inline server/services/events.js]: please paginate this

File: server/utils/community.js [hunk 2]
```diff
@@ -61,44 +54,9 @@ utils.addNewCommunity = async (data) => {
...
-  if (error) {
-    if (error.name === 'MongoError' && error.code === 11000) {
-      return {data: metadataObj}
-    }
-    throw error
-  }
-  return {data: metadataObj}
+utils.getLastBlockNumber = async () => {
+  const communityObj = await community.getModel().find().sort({blockNumber: -1}).limit(1)
+  return communityObj.length ? communityObj[0].blockNumber : 0
```
Comment (bejavu) [inline server/utils/community.js]: if not exist, should not be zero but the block of deployment.

File: server/utils/events.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  const blockNumber = event.blockNumber
+  console.log(`recieved TokenCreated event at ${blockNumber} blockNumber`)
+  const eventArgs = event.returnValues
+  const ccAddress = eventArgs.token
+  const owner = eventArgs.owner
+  const factoryAddress = event.address
+  const communityData = await utils.getCommunityData(factoryAddress, ccAddress)
+
+  try {
+    const community = await utils.addNewCommunity({ccAddress, owner, blockNumber, ...communityData})
```
Comment (bejavu) [inline server/utils/events.js]: this should use upsert if you are using in in several places

File: server/utils/community.js [hunk 4]
```diff
@@ -54,6 +54,11 @@ utils.addNewCommunity = async (data) => {
   return community.create(data)
 }
 
+utils.upsertCommunity = async (data) => {
```
Comment (bejavu) [inline server/utils/community.js]: db operations (like upsert) should be in the model

**Comment by leonprou** [comment]
@bejavu pagination was rather slow, cause you need to do `numberOfNewBlocks / pageSize` queries. So I added it as an option in case of error.

**Pull request merged**